tests/contrib/psycopg/test_psycopg.py

```diff
@@ -408,11 +413,34 @@ def test_postgres_dbm_propagation_comment(self):
         conn = psycopg2.connect(**POSTGRES_CONFIG)
         cursor = conn.cursor()
         cursor.__wrapped__ = mock.Mock()
+        # test string queries
         cursor.execute("select 'blah'")
         cursor.executemany("select %s", (("foo",), ("bar",)))
         dbm_comment = " /*dddbs='postgres',dde='staging',ddps='orders-app',ddpv='v7343437-d7ac743'*/"
         cursor.__wrapped__.execute.assert_called_once_with(dbm_comment + "select 'blah'")
         cursor.__wrapped__.executemany.assert_called_once_with(dbm_comment + "select %s", (("foo",), ("bar",)))
+        # test composed queries
+        cursor.__wrapped__.reset_mock()
+        cursor.execute(SQL("select 'blah'"))
+        cursor.executemany(SQL("select %s"), (("foo",), ("bar",)))
+        dbm_comment = " /*dddbs='postgres',dde='staging',ddps='orders-app',ddpv='v7343437-d7ac743'*/"
+        cursor.__wrapped__.execute.assert_called_once_with(
+            Composed(
+                [
+                    SQL(" /*dddbs='postgres',dde='staging',ddps='orders-app',ddpv='v7343437-d7ac743'*/"),
+                    SQL("select 'blah'"),
+                ]
+            )
+        )
```

Comment:
Do we have a test that when dbm is disabled that this is `Composed([SQL("select 'blah'")])` ?

I want to be sure the data type passed to psycopg2 is the exact same (`Composed` and not `SQL`)

Reply:
Yup there's this preexisting test: https://github.com/DataDog/dd-trace-py/blob/v1.6.2/tests/contrib/psycopg/test_psycopg.py#L267. Here the `DD_DBM_PROPAGATION_MODE` is not set so the sql statement is not changed.

Reply:
I am not sure that tests what I expect/want it to test. Or at least it doesn't assert it.

That is testing that with a `SQL()` that we still properly trace and collect the right resource name.

But we are passing a `SQL()` (correct?) into `cur.execute()`. With the DBM stuff that is being changed to a `Composed()` correct? Is that a problem? Would/could psycopg2 behavior differently if we are passing in a different type than the user provided?

Reply:
We pass in a SQL() object in the test. In the DBM logic we generate another SQL() object containing the dbm comment and we add those two SQL objects together. Adding two SQL() objects produces the composed object in the assertion. According to the psycopg spec creating a composed object from two SQL object is the preferred way of modifying a query: https://www.psycopg.org/docs/sql.html#sql-objects.

ex: `SQL("/* dbm comment */") + SQL("/* select *; */") -> Composed([SQL("/* dbm comment */"), SQL("/* select *; */"])`

Both SQL() and Composed() implement the Composable interface. But you're right they're not the same type.
`isinstance(psycopg.sql.SQL('h')), psycopg.sql.Composable)== True`
`isinstance(psycopg.sql.Composed([psycopg.sql.SQL('h'), psycopg.sql.SQL('h')]), psycopg.sql.Composable)== True`
`isinstance(psycopg.sql.SQL('h'), psycopg.sql.Composed)== False`

To answer your questions:
1. Yes,  this logic converts SQL() objects to Composed(). But if a composed object is passed to `cursor.execute()` then we don't change the type we simply append to it. 
2. This behavior is expected and shouldn't cause any unexpected side effects.

Alternative approach:

-  SQL() is passed to cursor.execute() get the raw string and create a new SQL() object with the DBM object
- If Composed() is passed to cursor.execute() create a SQL() object with the DBM comment and append it to the existing Composed. 

The con with this alternative approach is that it goes against psycopg composable conventions.


@brettlangdon  do you have any strong feelings? We can move forward with the alternative approach and put this PR on hold for now.

Reply:
Ok, I don't think I understood the class hierarchy with `Composed`/`SQL` and `Composable`.

Given that they are both `Composable`... and the preferred way from psycopg2, then seems fine to me.


My concern was if we changed a `str` to an `int` type of change, which this doesn't seem to be.